### PR TITLE
CBMs and NPCs

### LINKED ISSUE #24474: [RFC] NPCs and CBMs: upgrading your allies
The autodoc should have an option to install or remove CBMs from an NPC ally. install_bionics is a member of the player class, so it should apply to NPCs already.

The iexamine::autodoc function would need to detect additional friendly NPCs in the vicinity, and if there are any, query whether to install the CBMs on the player or one of the NPCs. I think most of the rest of the logic flows pretty much unchanged, replacing player &p with &player selected_NPC in a couple of instances.

Obviously, NPCs won't have the AI to use active CBMs, but the passive stat boosters and armor could be applied to them right now. And a future revision of the NPC AI code could let them take advantage of some of the active defense CBMs like dielectric capacitance or active utility CBMs like aero-evaporator.

I intend to implement this sometime in the next two weeks unless someone tells me not to or wants to take it over themselves.

### LINKED ISSUE #24475: [RFC] NPCs and CBMs: Allies upgrading you
If the PC has an NPC ally with superior skills, the player should have the option of asking the NPC to install CBMs through an autodoc.

The iexamine::autodoc function would need to be revised to detect NPCs in the area, calculate their effective skill for installing CBMs, and give the option of letting the NPC do the install instead of the player if the NPC's skill is better.

I intend to implement this sometime in the next few weeks unless someone tells me not to or someone else wants to implement in earlier themselves.

### PR DESCRIPTION
**Pull request by mlangsdorf**
Allow skilled, allied NPCs to install CBMs if they would have a higher chance of success than the survivor.

Allow CBMs with the BIONICS_NPC_USABLE flag to be installed on NPCs.

Fixes #24474 
Fixes #24475

### CODE REVIEW

**Comment by mlangsdorf** [comment]
Revised best installer images: Wilmer Tinsley knows what to do, but is sedated from the first operation. The player is warned that the player doesn't really know what he's doing.
better_installer_is_asleep1
The player persists, but Forest Madrid is more skilled and takes over.
better_installer_is_asleep2

File: src/iexamine.cpp [hunk 1]
```diff
@@ -3535,45 +3596,57 @@ void iexamine::autodoc( player &p, const tripoint &examp )
             const itype *itemtype = it->type;
             const bionic_id &bid = itemtype->bionic->id;
 
-            if( p.has_bionic( bid ) ) {
-                popup( _( "You have already installed this bionic."  ) );
+            if( patient.is_npc() && !bid->npc_usable ) {
+                //~ %2$s is the bionic CBM display name, %2$s is the patient name
+                popup( _( "%1$s cannot be installed on npc %2$s"), it->display_name().c_str(), patient.name );
```
Comment (DracoGriffin) [inline src/iexamine.cpp]: Suggest either:
removing `npc`
or
changing to: `cannot be installed on follower %2$s` as currently non-followers (hostile, neutral, or followers "on guard duty"/assigned to base) can't be operated on.

First option may be more prudent for future development (operations may take place on captured NPCs or whatever else)

File: src/npctalk.cpp [hunk 2]
```diff
@@ -654,6 +659,10 @@ std::string dialogue::dynamic_line( const talk_topic &the_topic ) const
                    _( "&You are deaf and can't talk. When you don't respond, %s becomes angry!" ),
                    beta->name.c_str() );
     }
+    if( topic == "TALK_SEDATED" ) {
+        return string_format( _( "%s is sedated and can't wake up until the sedation wears off" ),
```
Comment (DracoGriffin) [inline src/npctalk.cpp]: Suggest changing to `%s is sedated and cannot be moved or woken up until the [medication/sedation] wears off.`

**Comment by DracoGriffin** [comment]
Here is the mock-up I was thinking of for deciding on NPCs to installation:

mockupchoice

In this case, "unavailable" could be for whatever catch-all reason: sedated, missing, out of sight, etc.

Additionally, it could just default on the highest success chance (Roscoe David) to reduce keypresses, but in the case the NPC is unavailable, it would default on next highest (Elayne Solomon).

The `"WARNING: %i percent chance of genetic damage, blood loss, or damage to existing bionics! Continue anyway?` could either be another prompt after, or it could be at the bottom of that menu that changes based on which member is selected.

**Comment by Leland** [comment]
> it could just default on the highest success chance

I can't think of a reason why a user would want to use an NPC with a lower chance of success – OP's existing UI makes the most sense to me (automatically using NPC w/highest success rate)

File: src/bionics.cpp [hunk 3]
```diff
@@ -857,12 +868,13 @@ bool player::uninstall_bionic( bionic_id const &b_id, int skill_level, bool auto
...
 
     if( !has_bionic( b_id ) ) {
-        popup( _( "You don't have this bionic installed." ) );
+        popup( _( "%s don't have this bionic installed." ), disp_name() );
         return false;
     }
 
     if( b_id == "bio_blaster" ) {
-        popup( _( "Removing your Fusion Blaster Arm would leave you with a useless stump." ) );
+        popup( _( "Removing %s Fusion Blaster Arm would leave %s with a useless stump." ),
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/bionics.cpp]: I thought disp_name( true ) returned <npcname>'s .

File: src/iexamine.cpp [hunk 4]
```diff
@@ -3475,6 +3479,74 @@ void iexamine::climb_down( player &p, const tripoint &examp )
...
+    } );
+    int player_cos = bionic_manip_cos( player_skill, true, difficulty );
+    for( size_t i = 0; i < g->allies().size() ; i ++ ) {
+        if( ally_skills[ i ].first > player_skill ) {
+            const npc *e = g->allies()[ ally_skills[ i ].second ];
+            player &ally = *g->critter_by_id<player>( e->getID() );
+            int ally_cos = bionic_manip_cos( ally_skills[ i ].first, true, difficulty );
+            if( e->has_effect( effect_sleep ) ) {
+                //~ %1$s is the name of the ally
+                if( !g->u.query_yn( string_format( _( "%1$s is asleep, but has a %2$d chance of success compared to your %3$d chance of success.  Continue with a higher risk of failure?" ), ally.disp_name(), ally_cos, player_cos ) ) ) {
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/iexamine.cpp]: afaict, popups don't take color parameters.

File: src/iexamine.cpp [hunk 5]
```diff
@@ -3521,6 +3594,13 @@ void iexamine::autodoc( player &p, const tripoint &examp )
 
     amenu.query();
 
+    auto popup_player_or_npc = []( player &pat, const std::string player_str, const std::string npc_str ) {
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/iexamine.cpp]: Fixed.

File: src/iexamine.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/iexamine.cpp]: nope, I was wrong, you just have to specify the color and can't use good/bad.

File: src/bionics.cpp [hunk 6]
```diff
@@ -1106,10 +1074,10 @@ void bionics_install_failure( player *u, int difficulty, int success, bool autod
 
     switch( rng( 1, 5 ) ) {
         case 1:
-            add_msg( m_neutral, _( "You flub the installation." ) );
+            add_msg( m_neutral, _( "%s flub the installation." ), installer.disp_name() );
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/bionics.cpp]: Fixed.

File: src/iexamine.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/iexamine.cpp]: Fixed.

File: src/bionics.cpp [hunk 7]
```diff
@@ -1000,32 +997,33 @@ bool player::install_bionics( const itype &type, int skill_level, bool autodoc )
...
     } else {
-        if( !query_yn(
+        if( !g->u.query_yn(
                 _( "WARNING: %i percent chance of genetic damage, blood loss, or damage to existing bionics! Continue anyway?" ),
                 ( 100 - int( chance_of_success ) ) ) ) {
             return false;
         }
     }
 
     // Practice skills only if conducting manual self-installation
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/bionics.cpp]: Fixed, thanks.

File: src/bionics.cpp [hunk 8]
```diff
@@ -776,14 +774,14 @@ void player::process_bionic( int b )
     }
 }
 
-void bionics_uninstall_failure( player *u )
+void player::bionics_uninstall_failure( std::string installer_name )
 {
     switch( rng( 1, 5 ) ) {
         case 1:
-            add_msg( m_neutral, _( "You flub the removal." ) );
+            add_msg( m_neutral, _( "%s flub the removal." ), installer_name );
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/bionics.cpp]: Fixed.

File: src/npctalk.cpp (hunk 2 above)
Comment (mlangsdorf) [reply src/npctalk.cpp]: Fixed.

File: src/iexamine.cpp (hunk 1 above)
Comment (mlangsdorf) [reply src/iexamine.cpp]: Fixed.

**Comment by mlangsdorf** [comment]
I would really rather not have to build a complicated menu, as opposed to the current easy solution of just using the best installer and warning if the best installer isn't available.

File: src/bionics.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Night-Pryanik) [reply src/bionics.cpp]: Ок.

File: src/iexamine.cpp [hunk 9]
```diff
@@ -3616,8 +3609,8 @@ void iexamine::autodoc( player &p, const tripoint &examp )
             const bionic_id &bid = itemtype->bionic->id;
 
             if( patient.is_npc() && !bid->npc_usable ) {
-                //~ %2$s is the bionic CBM display name, %2$s is the patient name
-                popup( _( "%1$s cannot be installed on npc %2$s"), it->display_name().c_str(), patient.name );
+                //~ %1$s is the bionic CBM display name, %2$s is the patient name
+                popup( _( "%1$s cannot be installed on %2$s"), it->display_name().c_str(), patient.name );
```
Comment (Night-Pryanik) [inline src/iexamine.cpp]: Period in the end.

File: src/npctalk.cpp [hunk 10]
```diff
@@ -660,7 +660,7 @@ std::string dialogue::dynamic_line( const talk_topic &the_topic ) const
                    beta->name.c_str() );
     }
     if( topic == "TALK_SEDATED" ) {
-        return string_format( _( "%s is sedated and can't wake up until the sedation wears off" ),
+        return string_format( _( "%s is sedated and can't be moved or woken up until the medication or sedation wears off" ),
```
Comment (Night-Pryanik) [inline src/npctalk.cpp]: Period in the end.

Comment (mlangsdorf) [reply src/npctalk.cpp]: Fixed, thanks.

File: src/iexamine.cpp (hunk 9 above)
Comment (mlangsdorf) [reply src/iexamine.cpp]: Fixed, thanks.

**Comment by mlangsdorf** [comment]
I can't figure out the astyle issue that's stopping the gorgon build, so if someone else can, pointers are appreciated.

**Comment by AMurkin** [comment]
Yours and astyled versions diff of output.h:
```diff
@@ -19,8 +19,8 @@
 struct input_event;
 struct iteminfo;
 enum direction : unsigned;
-    class input_context;
-    namespace units
+class input_context;
+namespace units
 {
 template<typename V, typename U>
 class quantity;
@@ -407,7 +407,7 @@ inline void full_screen_popup( const char *mes, Args &&... args )
     popup( string_format( mes, std::forward<Args>( args )... ), PF_FULLSCREEN );
 }
 template<typename ...Args>
-inline void popup_player_or_npc( player &p, const char *player_mes , const char *npc_mes,
+inline void popup_player_or_npc( player &p, const char *player_mes, const char *npc_mes,
                                  Args &&... args )
 {
     if( p.is_player() ) {
```

**Comment by mlangsdorf** [comment]
Fixed. Don't know what was going on there.

**Comment by mlangsdorf** [comment]
I'll figure out the gcc compile issue tomorrow.

**Pull request merged**